File: narrative_os/schemas/benchmark.py
"""schemas/benchmark.py — Benchmark Phase 1 请求/响应模型。"""
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator

from narrative_os.schemas.traces import Run


class BenchmarkJobType(str, Enum):
    PROJECT_BENCHMARK = "project_benchmark"
    AUTHOR_DISTILLATION = "author_distillation"


class BenchmarkJobMode(str, Enum):
    SINGLE_WORK = "single_work"
    MULTI_WORK = "multi_work"
    SCENE = "scene"


class BenchmarkSourceType(str, Enum):
    PROJECT_REFERENCE = "project_reference"
    AUTHOR_REFERENCE = "author_reference"


class BenchmarkProfileStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"


class BenchmarkSkillApplyMode(str, Enum):
    GUIDE = "guide"
    HYBRID = "hybrid"
    STRICT = "strict"


class BenchmarkSceneType(str, Enum):
    BATTLE = "battle"
    EMOTION = "emotion"
    REVEAL = "reveal"
    DAILY = "daily"
    ENSEMBLE = "ensemble"
    GENERAL = "general"


class BenchmarkSourceCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    file_name: str = Field(default="", max_length=300)
    text: str = Field(default="")
    author_name: str | None = Field(default=None, max_length=200)
    corpus_group: str = Field(default="", max_length=100)
    chapter_sep: str | None = Field(default=None, max_length=300)


class BenchmarkJobCreateRequest(BaseModel):
    job_type: BenchmarkJobType
    mode: BenchmarkJobMode = BenchmarkJobMode.SINGLE_WORK
    source_ids: list[str] = Field(default_factory=list)
    sources: list[BenchmarkSourceCreateRequest] = Field(default_factory=list)
    chapter_sep: str | None = Field(default=None, max_length=300)
    extract_snippets: bool = True
    target_platform: str | None = Field(default=None, max_length=100)
    author_name: str | None = Field(default=None, max_length=200)
    corpus_group: str | None = Field(default=None, max_length=100)

    @model_validator(mode="after")
    def _validate_sources(self) -> "BenchmarkJobCreateRequest":
        if not self.source_ids and not self.sources:
            raise ValueError("source_ids 或 sources 至少提供一项。")
        if self.job_type == BenchmarkJobType.AUTHOR_DISTILLATION and not (
            self.author_name or self.corpus_group
        ):
            raise ValueError("author_distillation 必须提供 author_name 或 corpus_group。")
        return self


class BenchmarkSource(BaseModel):
    source_id: str
    project_id: str
    corpus_group: str = ""
    title: str
    author_name: str | None = None
    file_name: str = ""
    sha256: str = ""
    char_count: int = 0
    chapter_sep: str | None = None
    source_type: BenchmarkSourceType
    created_at: str


class BenchmarkSnippet(BaseModel):
    snippet_id: str
    profile_id: str
    project_id: str
    source_id: str
    scene_type: BenchmarkSceneType
    chapter: int | None = None
    offset_start: int = 0
    offset_end: int = 0
    char_count: int = 0
    anchor_score: float = 0.0
    purity_score: float = 0.0
    distinctiveness_score: float = 0.0
    source_hit_verified: bool = False
    text: str


class BenchmarkProfile(BaseModel):
    profile_id: str
    project_id: str
    profile_type: BenchmarkJobType
    profile_name: str = ""
    source_ids: list[str] = Field(default_factory=list)
    stable_traits: list[dict[str, Any]] = Field(default_factory=list)
    conditional_traits: list[dict[str, Any]] = Field(default_factory=list)
    anti_traits: list[dict[str, Any]] = Field(default_factory=list)
    scene_anchors: dict[str, Any] = Field(default_factory=dict)
    humanness_baseline: dict[str, Any] = Field(default_factory=dict)
    status: BenchmarkProfileStatus = BenchmarkProfileStatus.DRAFT
    snippet_count: int = 0
    created_at: str
    activated_at: str | None = None


class AuthorSkillProfile(BaseModel):
    skill_id: str
    origin_project_id: str
    run_id: str | None = None
    skill_name: str = ""
    author_name: str = ""
    source_ids: list[str] = Field(default_factory=list)
    stable_rules: list[dict[str, Any]] = Field(default_factory=list)
    conditional_rules: list[dict[str, Any]] = Field(default_factory=list)
    anti_rules: list[dict[str, Any]] = Field(default_factory=list)
    dialogue_rules: list[dict[str, Any]] = Field(default_factory=list)
    scene_patterns: dict[str, Any] = Field(default_factory=dict)
    chapter_hook_patterns: list[dict[str, Any]] = Field(default_factory=list)
    humanness_baseline: dict[str, Any] = Field(default_factory=dict)
    confidence_map: dict[str, Any] = Field(default_factory=dict)
    status: BenchmarkProfileStatus = BenchmarkProfileStatus.DRAFT
    created_at: str
    applied: bool = False
    application_mode: BenchmarkSkillApplyMode | None = None


class BenchmarkSkillListResponse(BaseModel):
    items: list[AuthorSkillProfile] = Field(default_factory=list)
    active_skill_id: str | None = None
    active_mode: BenchmarkSkillApplyMode | None = None


class BenchmarkSkillApplyRequest(BaseModel):
    mode: BenchmarkSkillApplyMode = BenchmarkSkillApplyMode.HYBRID


class BenchmarkSkillApplyResponse(BaseModel):
    skill: AuthorSkillProfile
    mode: BenchmarkSkillApplyMode
    message: str = ""


class BenchmarkJobCreateResponse(BaseModel):
    run_id: str
    status: str
    profile: BenchmarkProfile
    author_skill: AuthorSkillProfile | None = None
    source_ids: list[str] = Field(default_factory=list)
    snippet_count: int = 0
    message: str = ""


class BenchmarkJobDetailResponse(BaseModel):
    run: Run
    profile: BenchmarkProfile | None = None
    author_skill: AuthorSkillProfile | None = None
    sources: list[BenchmarkSource] = Field(default_factory=list)
    snippets: list[BenchmarkSnippet] = Field(default_factory=list)
    message: str = ""


class BenchmarkProfileActivateRequest(BaseModel):
    profile_id: str = Field(..., min_length=1)


class BenchmarkSnippetListResponse(BaseModel):
    profile: BenchmarkProfile | None = None
    items: list[BenchmarkSnippet] = Field(default_factory=list)


class BenchmarkScore(BaseModel):
    score_id: str
    project_id: str
    run_id: str | None = None
    chapter: int = 0
    profile_id: str
    humanness_score: float = 0.0
    adherence_score: float = 0.0
    dimension_scores: dict[str, float] = Field(default_factory=dict)
    violations: list[str] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    created_at: str

File: frontend/src/api/benchmark.ts
import client from './client'

export type BenchmarkSceneType = 'battle' | 'emotion' | 'reveal' | 'daily' | 'ensemble' | 'general'
export type BenchmarkJobType = 'project_benchmark' | 'author_distillation'
export type BenchmarkSkillApplyMode = 'guide' | 'hybrid' | 'strict'

export interface BenchmarkSourceInput {
  title: string
  file_name?: string
  text: string
  author_name?: string | null
  corpus_group?: string
  chapter_sep?: string | null
}

export interface BenchmarkTrait {
  dimension: string
  name: string
  summary: string
  value?: number
}

export interface BenchmarkProfile {
  profile_id: string
  project_id: string
  profile_type: 'project_benchmark' | 'author_distillation'
  profile_name: string
  source_ids: string[]
  stable_traits: BenchmarkTrait[]
  conditional_traits: BenchmarkTrait[]
  anti_traits: BenchmarkTrait[]
  scene_anchors: Record<string, Array<Record<string, unknown>>>
  humanness_baseline: Record<string, unknown>
  status: 'draft' | 'active'
  snippet_count: number
  created_at: string
  activated_at?: string | null
}

export interface AuthorSkillProfile {
  skill_id: string
  origin_project_id: string
  run_id?: string | null
  skill_name: string
  author_name: string
  source_ids: string[]
  stable_rules: BenchmarkTrait[]
  conditional_rules: BenchmarkTrait[]
  anti_rules: BenchmarkTrait[]
  dialogue_rules: BenchmarkTrait[]
  scene_patterns: Record<string, Array<Record<string, unknown>>>
  chapter_hook_patterns: Array<Record<string, unknown>>
  humanness_baseline: Record<string, unknown>
  confidence_map: Record<string, unknown>
  status: 'draft' | 'active'
  created_at: string
  applied: boolean
  application_mode?: BenchmarkSkillApplyMode | null
}

export interface BenchmarkSkillListResponse {
  items: AuthorSkillProfile[]
  active_skill_id?: string | null
  active_mode?: BenchmarkSkillApplyMode | null
}

export interface BenchmarkSkillApplyResponse {
  skill: AuthorSkillProfile
  mode: BenchmarkSkillApplyMode
  message: string
}

export interface BenchmarkSnippet {
  snippet_id: string
  profile_id: string
  project_id: string
  source_id: string
  scene_type: BenchmarkSceneType
  chapter?: number | null
  offset_start: number
  offset_end: number
  char_count: number
  anchor_score: number
  purity_score: number
  distinctiveness_score: number
  source_hit_verified: boolean
  text: string
}

export interface BenchmarkSource {
  source_id: string
  project_id: string
  corpus_group: string
  title: string
  author_name?: string | null
  file_name: string
  sha256: string
  char_count: number
  chapter_sep?: string | null
  source_type: 'project_reference' | 'author_reference'
  created_at: string
}

export interface BenchmarkJobCreateRequest {
  job_type: BenchmarkJobType
  mode: 'single_work' | 'multi_work' | 'scene'
  source_ids?: string[]
  sources: BenchmarkSourceInput[]
  chapter_sep?: string | null
  extract_snippets?: boolean
  target_platform?: string | null
  author_name?: string | null
  corpus_group?: string | null
}

export interface BenchmarkJobCreateResponse {
  run_id: string
  status: string
  profile: BenchmarkProfile
  author_skill?: AuthorSkillProfile | null
  source_ids: string[]
  snippet_count: number
  message: string
}

export interface BenchmarkJobDetailResponse {
  run: {
    run_id: string
    run_type: string
    status: string
    steps: Array<{
      step_id: string
      step_index: number
      agent_name: string
      status: string
      artifact?: {
        artifact_type: string
        output_content: string
      } | null
    }>
  }
  profile?: BenchmarkProfile | null
  author_skill?: AuthorSkillProfile | null
  sources: BenchmarkSource[]
  snippets: BenchmarkSnippet[]
  message: string
}

export interface BenchmarkSnippetListResponse {
  profile?: BenchmarkProfile | null
  items: BenchmarkSnippet[]
}

export const benchmark = {
  createJob: (id: string, data: BenchmarkJobCreateRequest) =>
    client.post<BenchmarkJobCreateResponse>(`/projects/${id}/benchmark/jobs`, data),
  getJob: (id: string, runId: string) =>
    client.get<BenchmarkJobDetailResponse>(`/projects/${id}/benchmark/jobs/${runId}`),
  getProfile: (id: string) =>
    client.get<BenchmarkProfile | null>(`/projects/${id}/benchmark/profile`),
  listSnippets: (id: string, params?: { profile_id?: string; scene_type?: string; limit?: number }) =>
    client.get<BenchmarkSnippetListResponse>(`/projects/${id}/benchmark/snippets`, { params }),
  activateProfile: (id: string, profileId: string) =>
    client.post<BenchmarkProfile>(`/projects/${id}/benchmark/profile/activate`, { profile_id: profileId }),
  listSkills: (id: string, params?: { author_name?: string; limit?: number }) =>
    client.get<BenchmarkSkillListResponse>(`/projects/${id}/benchmark/skills`, { params }),
  getSkill: (id: string, skillId: string) =>
    client.get<AuthorSkillProfile>(`/projects/${id}/benchmark/skills/${skillId}`),
  applySkill: (id: string, skillId: string, mode: BenchmarkSkillApplyMode) =>
    client.post<BenchmarkSkillApplyResponse>(`/projects/${id}/benchmark/skills/${skillId}/apply`, { mode }),
}